### PR TITLE
Fix MemBehavior and TempRValueOpt to handle access markers

### PR DESCRIPTION
**Pull request by atrick**
The first commit cleans up the infrastructure and improves MemBehavior.

The second commit fixes the TempRValue pass to correctly handle access markers.

Both commits are required together for correct behavior, so I'm smashing them into one PR.

### CODE REVIEW

**Comment by atrick** [comment]
@swift-ci test

**Comment by atrick** [comment]
@swift-ci benchmark

**Comment by atrick** [comment]
@swift-ci test source compatibility

**Comment by atrick** [comment]
@swift-ci test

**Comment by atrick** [comment]
@swift-ci benchmark

**Comment by atrick** [comment]
@swift-ci test source compatibility

File: lib/SILOptimizer/Transforms/TempRValueElimination.cpp [hunk 1]
```diff
@@ -474,33 +487,48 @@ void TempRValueOptPass::run() {
   bool changed = false;
 
   // Find all copy_addr instructions.
+  llvm::SmallPtrSet<CopyAddrInst *, 8> deadCopies;
```
Comment (gottesmm) [inline lib/SILOptimizer/Transforms/TempRValueElimination.cpp]: Can you use a small vector here? If you are worried about multiple dead copies, can you just do a sortUnique on a small vector?

File: lib/SIL/MemAccessUtils.cpp [hunk 2]
```diff
@@ -21,6 +21,56 @@
...
+    case ValueKind::TailAddrInst:
+    case ValueKind::IndexAddrInst:
+      v = cast<SingleValueInstruction>(v)->getOperand(0);
+      continue;
+    };
+  }
+}
+
+bool swift::isLetAddress(SILValue accessedAddress) {
+  assert(accessedAddress == getAccessedAddress(accessedAddress)
```
Comment (gottesmm) [inline lib/SIL/MemAccessUtils.cpp]: Why not just do this for the user rather than assert it?

File: lib/SIL/MemAccessUtils.cpp [hunk 3]
```diff
@@ -21,6 +21,56 @@
...
+         && "caller must find the address root");
+  // Is this an address of a "let" class member?
+  if (auto *REA = dyn_cast<RefElementAddrInst>(accessedAddress))
```
Comment (gottesmm) [inline lib/SIL/MemAccessUtils.cpp]: Fix names here?

File: include/swift/SILOptimizer/Analysis/ValueTracking.h [hunk 4]
```diff
@@ -50,7 +50,8 @@ bool pointsToLocalObject(SILValue V);
 /// indirection (e.g. ref_element_addr, project_box, etc.).
 inline bool isUniquelyIdentified(SILValue V) {
   return pointsToLocalObject(V)
-         || isExclusiveArgument(getUnderlyingAddressRoot(V));
+         || (V->getType().isAddress()
```
Comment (gottesmm) [inline include/swift/SILOptimizer/Analysis/ValueTracking.h]: Was this a bug? Do we have a test case for this?

File: include/swift/SIL/SILNodes.def [hunk 5]
```diff
@@ -591,6 +591,9 @@ ABSTRACT_VALUE_AND_INST(SingleValueInstruction, ValueBase, SILInstruction)
                     SingleValueInstruction, MayHaveSideEffects, DoesNotRelease)
   SINGLE_VALUE_INST(StoreBorrowInst, store_borrow,
                     SILInstruction, MayWrite, DoesNotRelease)
+  // begin_access may trap. Trapping is unordered with respect to memory access,
+  // and with respect to other traps, but it is still conservatively considered
```
Comment (gottesmm) [inline include/swift/SIL/SILNodes.def]: Nice!

File: include/swift/SIL/SILNodes.def [hunk 6]
```diff
@@ -803,6 +806,10 @@ NON_VALUE_INST(DestroyValueInst, destroy_value,
                SILInstruction, MayHaveSideEffects, MayRelease)
 NON_VALUE_INST(EndBorrowInst, end_borrow,
                SILInstruction, MayHaveSideEffects, DoesNotRelease)
+// end_access is considered to have side effects because it modifies runtime
+// state outside of the memory modified by the access that is visible to
```
Comment (gottesmm) [inline include/swift/SIL/SILNodes.def]: Nice!

File: include/swift/SIL/MemAccessUtils.h [hunk 7]
```diff
@@ -46,7 +46,47 @@
...
+/// assumed to store directly to the same, unaliased accessedAddress.
+///
+/// The address of a let-variable must be a formal access root. A 'let' member
+/// of a struct is *not* a let-variable, because it's memory may be written when
+/// formally modifying the outer struct. A let-variable is either an entire
+/// local variable, global variable, or class property (this is the definition
+/// of an address root).
+///
+/// The caller should derive the accessed address using
+/// stripAccessMarkers(getAccessedAddress(ptr)).
```
Comment (gottesmm) [inline include/swift/SIL/MemAccessUtils.h]: If you are going to assert this internal to this function in the implementation, please say that the API has such a precondition in the interface explicitly!

File: lib/IRGen/IRGenSIL.cpp [hunk 8]
```diff
@@ -3567,8 +3568,11 @@ void IRGenSILFunction::visitRefTailAddrInst(RefTailAddrInst *i) {
 }
 
 static bool isInvariantAddress(SILValue v) {
-  auto root = getUnderlyingAddressRoot(v);
-  if (auto ptrRoot = dyn_cast<PointerToAddressInst>(root)) {
+  SILValue accessedAddress = getAccessedAddress(v);
+  if (accessedAddress->getType().isAddress() && isLetAddress(accessedAddress)) {
```
Comment (gottesmm) [inline lib/IRGen/IRGenSIL.cpp]: I have been noticing the Let Address code interspersed through out. Was this just something that you noticed? Shouldn't it be in a different commit?

File: lib/IRGen/IRGenSIL.cpp [hunk 9]
```diff
@@ -3567,8 +3568,11 @@ void IRGenSILFunction::visitRefTailAddrInst(RefTailAddrInst *i) {
...
+    return true;
+  }
+  if (auto ptrRoot = dyn_cast<PointerToAddressInst>(accessedAddress)) {
```
Comment (gottesmm) [inline lib/IRGen/IRGenSIL.cpp]: `auto *`

File: lib/SIL/MemAccessUtils.cpp [hunk 10]
```diff
@@ -21,6 +21,56 @@
 
 using namespace swift;
 
+SILValue swift::stripAccessMarkers(SILValue V) {
```
Comment (gottesmm) [inline lib/SIL/MemAccessUtils.cpp]: `V` -> `v`

File: lib/SILOptimizer/Analysis/MemoryBehavior.cpp [hunk 11]
```diff
@@ -76,23 +107,55 @@ class MemoryBehaviorVisitor
     // If we do not have any more information, just use the general memory
     // behavior implementation.
     auto Behavior = Inst->getMemoryBehavior();
-    bool ReadOnlyAccess = isLetPointer(V);
```
Comment (gottesmm) [inline lib/SILOptimizer/Analysis/MemoryBehavior.cpp]: Nice! Much better!

File: lib/SILOptimizer/Transforms/TempRValueElimination.cpp [hunk 12]
```diff
@@ -336,8 +342,15 @@ bool TempRValueOptPass::tryOptimizeCopyIntoTemp(CopyAddrInst *copyInst) {
...
     return false;
 
-  assert(tempObj != copyInst->getSrc() &&
-         "can't initialize temporary with itself");
+  // The copy's source address must not be a scoped instruction. When the
+  // temporary object is eliminated, it's uses are replaced with the copy's
+  // source. Therefore, the source address must be valid at least until the next
+  // instruction that may write to or destroy the source. End-of-scope markers
+  // do not write to or destroy memory, so scoped addresses are not valid
+  // replacements.
```
Comment (gottesmm) [inline lib/SILOptimizer/Transforms/TempRValueElimination.cpp]: Can you provide an example here or define what is a "scoped" instruction. It is undefined when I pretend to not know what you probably mean.

File: lib/SILOptimizer/Transforms/TempRValueElimination.cpp (hunk 1 above)
Comment (atrick) [reply lib/SILOptimizer/Transforms/TempRValueElimination.cpp]: Yes, the same copy can be inserted multiple times, which I was trying to communicate with a set. I can change it to sortUnique.

Comment (atrick) [reply lib/SILOptimizer/Transforms/TempRValueElimination.cpp]: Also, I was originally checking membership and deleting from the set, but I'm not doing that anymore.

File: lib/SIL/MemAccessUtils.cpp (hunk 2 above)
Comment (atrick) [reply lib/SIL/MemAccessUtils.cpp]: Because whoever uses this needs to be aware that they are using the correct address to represent a formal access, and use the same address for other comparisons and queries. We should not try to hide that. That's where I've been seeing bugs.

File: lib/SIL/MemAccessUtils.cpp (hunk 3 above)
Comment (atrick) [reply lib/SIL/MemAccessUtils.cpp]: I think you're talking about the remnants of capitalization. Sure I can fix that now.

File: include/swift/SILOptimizer/Analysis/ValueTracking.h (hunk 4 above)
Comment (atrick) [reply include/swift/SILOptimizer/Analysis/ValueTracking.h]: No bug, it was just assumed that we were working with an address. I want to make that explicit everywhere and added asserts. A major part of the confusion is that some memory APIs accept random values and other expect an address for a specific access.

File: lib/IRGen/IRGenSIL.cpp (hunk 8 above)
Comment (atrick) [reply lib/IRGen/IRGenSIL.cpp]: I just happened to rename it when I moved the utility and added the assert.

File: lib/IRGen/IRGenSIL.cpp (hunk 9 above)
Comment (atrick) [reply lib/IRGen/IRGenSIL.cpp]: Not part of this commit, but sure I'll fix it.

File: lib/SILOptimizer/Transforms/TempRValueElimination.cpp (hunk 12 above)
Comment (atrick) [reply lib/SILOptimizer/Transforms/TempRValueElimination.cpp]: Yes, good idea.

File: lib/SIL/MemAccessUtils.cpp (hunk 10 above)
Comment (atrick) [reply lib/SIL/MemAccessUtils.cpp]: Fixed

File: include/swift/SIL/MemAccessUtils.h (hunk 7 above)
Comment (atrick) [reply include/swift/SIL/MemAccessUtils.h]: I fixed up this comment

**Comment by atrick** [comment]
@swift-ci smoke test

**Comment by atrick** [comment]
@swift-ci smoke test

**Comment by atrick** [comment]
@swift-ci smoke test OS X

**Comment by atrick** [comment]
There was one significant benchmark regression (and many large improvements). I'm only working on correctness or deep fundamental optimization bugs now, not individual benchmark performance, but I'm tracking it here. 

Filed <rdar://59999424> Regression in Set.isStrict[Super|Sub]set.Seq.Box[0|25] benchmark after MemBehavior fixes
